### ISSUE TITLE
Simplif.simplify_lets generates incorrect code if fed its own output

### ISSUE THREAD

**Issue by lthls** [ocaml/ocaml#9954]
This is not a real bug, and more a lack of documentation and checks, but `Simplif.simplify_lets` will generate incorrect code on some seemingly correct lambda inputs. In particular, running the function on its own output can result in incorrect code being produced.

I've found two examples so far (sample ml programs and their lambda output after an extra round of `simplify_lets`).
Example 1:
```ocaml
let f x =
  let r = ref x in
  r := !r + 1;
  !r
```
compiles to:
```
(let (f/82 = (function x/84[int] : int (seq (assign r/85 (+ x/84 1)) x/84)))
```
The cause is [this optimisation](https://github.com/ocaml/ocaml/blob/b727f422bd48022a830566f19c7d5e50181a96e8/lambda/simplif.ml#L534), which replaces `let x = y in expr` by `expr[x <- y]` even for `Variable` bindings.

Example 2:
```ocaml
let g () =
  let r = ref 0 in
  let x = !r in
  (fun _ -> x)
```
compiles to:
```
(let g/86 =
     (function param/91
       (let (r/88 =v[int] 0) (function param/90 : int r/88)))))
```
which breaks the invariant that variables bound to a `Variable` let must not occur in nested function bodies (they can't be passed in the function's closure like regular variables, in general).

This was discovered while trying to force a different type for mutable variables in the lambda representation (and dedicated `Llet` and `Lvar` variants). I wasn't sure how to patch `simplify_lets` in these cases, and it turns out the current code is wrong for those cases where there are already some `Variable` bindings. Fortunately this cannot occur with the current code paths, as the only place where `Variable` bindings are generated is in `simplify_lets` itself, and the function doesn't call itself recursively on the result.

I'd still like to know if other maintainers are bothered by this behaviour and want to discuss how to fix it, or if it's just how it works and it's not worth spending too much time on it.

**Comment by gasche** [comment]
This indeed seems concerning to me, I would expect simplifications to be applicable at any point (potentially several times). At least it should be documented, and it would be nice to fix it.

**Comment by lthls** [comment]
I'll probably propose some changes to the representation of mutable variables in lambda at some point, which would include a fix for this, but that may not happen soon. I can review related PRs by other people though.

**Comment by gasche** [comment]
If I understand correctly, the problematic optimization translates

```ocaml
let y = x in e
```

into

```ocaml
e[x/y]
```

This is correct when y and x are usual variables, but fishy when either of them is actually an assignable (mutable variable). One worrying aspect of the current representation is that variable occurrences don't carry any metadata to tell whether they are mutable or not (we would have to use `Lmutvar` as in this proposal, or `Lvar of mut_flag * variable`); if the correctness of this optimization relied strongly on the mutability of `x`, we would be in trouble. (Some parts of Simplify carry information on which of the variables are mutable, but `count` does not, and it tries to guess what substitution would do, so it would be tricky to track this information there.)

However, when thinking about it, I believe that:

1. `let-mut x = y in e` should never be substituted away. This is easy to implement by checking the binding kind today.
2. For non-mutable bindings, `let<kind> x = mut y in e` can be *correctly* substituted away if the binding-kind information is correct, namely if it is `Alias`, or if it is `StrictOpt` and `x` does not occur in `e`. (It should not be `Strict` by definition.)

so it sounds like we should be able to fix the reported misoptimization by just disabling the substitution of mutable bindings.

**Comment by gasche** [comment]
If I understand correctly, this bug has been fixed by #10090.
